### PR TITLE
[SPARK-27813][SQL] DataSourceV2: Add DropTable logical operation

### PR DESCRIPTION
**Pull request by jzhuge**
## What changes were proposed in this pull request?

Support DROP TABLE from V2 catalogs.
Move DROP TABLE into catalyst.
Move parsing tests for DROP TABLE/VIEW to PlanResolutionSuite to validate existing behavior.
Add new tests fo catalyst parser suite.
Separate DROP VIEW into different code path from DROP TABLE.
Move DROP VIEW into catalyst as a new operator.
Add a meaningful exception to indicate view is not currently supported in v2 catalog.

## How was this patch tested?

New unit tests.
Existing unit tests in catalyst and sql core.

### CODE REVIEW

**Comment by jzhuge** [comment]
Split the SPARK-26946 commits into a separate PR https://github.com/apache/spark/pull/24689

**Comment by jzhuge** [comment]
Please test

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/IdentifierImpl.java [hunk 1]
```diff
@@ -49,6 +49,11 @@ public String name() {
     return name;
   }
 
+  @Override
+  public String toString() {
+    return Arrays.toString(namespace) + " " + name;
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/IdentifierImpl.java]: I think this should produce the quoted identifier that can be parsed, like what is produced by `CatalogV2Implicits.IdentifierHelper`.

**Comment by jzhuge** [comment]
Make a few changes

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -96,19 +95,13 @@ object AnalysisContext {
 class Analyzer(
     catalog: SessionCatalog,
     conf: SQLConf,
-    maxIterations: Int,
-    override val lookupCatalog: Option[(String) => CatalogPlugin] = None)
-  extends RuleExecutor[LogicalPlan] with CheckAnalysis with LookupCatalog {
+    maxIterations: Int)
+  extends RuleExecutor[LogicalPlan] with CheckAnalysis {
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I use this for `FunctionCatalog` in the rules that are defined as objects in the `Analyzer`. I'd rather keep it in. At a minimum, this change should be in a separate PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 3]
```diff
@@ -49,11 +50,19 @@ import org.apache.spark.util.random.RandomSampler
...
  * TableIdentifier.
  */
-class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging {
+class AstBuilder(conf: SQLConf)
+  extends SqlBaseBaseVisitor[AnyRef] with Logging with LookupCatalog {
   import ParserUtils._
 
   def this() = this(new SQLConf())
 
+  val defaultCatalog = for {
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: It is not obvious what this is doing. I prefer using `map` and `flatMap` for readability.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalog/v2/TestTableCatalog.scala [hunk 4]
```diff
@@ -89,6 +89,8 @@ class TestTableCatalog extends TableCatalog {
   }
 
   override def dropTable(ident: Identifier): Boolean = Option(tables.remove(ident)).isDefined
+
+  override def toString: String = s"${getClass.getSimpleName}($name)"
```
Comment (rdblue) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalog/v2/TestTableCatalog.scala]: Why include the class name?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 5]
```diff
@@ -640,11 +641,20 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
    * Create a [[DropTableCommand]] command.
    */
   override def visitDropTable(ctx: DropTableContext): LogicalPlan = withOrigin(ctx) {
-    DropTableCommand(
-      visitTableIdentifier(ctx.tableIdentifier),
-      ctx.EXISTS != null,
-      ctx.VIEW != null,
-      ctx.PURGE != null)
+    visitMultipartIdentifier(ctx.multipartIdentifier()) match {
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: The way we do this for `CreateTable` is to parse to an intermediate representation, `CreateTableStatement` that contains exactly what was parsed. Then translation to v1 `DropTableCommand` or v2 `DropTable` can be done in an analyzer rule. That way, the parser and the analyzer stay separate and the parser implementation for `DROP TABLE` can be kept in Catalyst.

I think this should take the same approach.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala [hunk 6]
```diff
@@ -162,4 +162,18 @@ class DataSourceV2SQLSuite extends QueryTest with SharedSQLContext with BeforeAn
       conf.setConfString("spark.sql.default.catalog", originalDefaultCatalog)
     }
   }
+
+  test("drop table from catalog") {
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: Nit: naming convention is `LogicalPlanName: case description`, like:
* `"DropTable: basic behavior"`
* `"DropTable: if exists"`
* `"DropTable: when table is missing"`

I think that these should be separate cases, instead of including the second `DROP TABLE` command below.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/LookupCatalog.scala [hunk 7]
```diff
@@ -26,27 +28,31 @@ import org.apache.spark.sql.catalyst.TableIdentifier
...
-        case Seq(catalogName, tail @ _*) =>
-          try {
-            val catalog = lookup(catalogName)
-            (Some(catalog), Identifier.of(tail.init.toArray, tail.last))
-          } catch {
-            case _: CatalogNotFoundException =>
-              (None, Identifier.of(parts.init.toArray, parts.last))
-          }
-      }
+    def unapply(parts: Seq[String]): Option[CatalogObjectIdentifier] = parts match {
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/LookupCatalog.scala]: Do these changes need to be done here, or can we do them in a separate commit?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (jzhuge) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Will keep it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/LookupCatalog.scala (hunk 7 above)
Comment (jzhuge) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/LookupCatalog.scala]: They are in the separate PR https://github.com/apache/spark/pull/24689.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 3 above)
Comment (jzhuge) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: removed.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalog/v2/TestTableCatalog.scala (hunk 4 above)
Comment (jzhuge) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalog/v2/TestTableCatalog.scala]: removed.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala (hunk 6 above)
Comment (jzhuge) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: renamed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/LookupCatalog.scala (hunk 7 above)
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/LookupCatalog.scala]: Can you remove these changes from this PR then?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/CatalogManager.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+import org.apache.spark.sql.internal.SQLConf;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+
+import static org.apache.spark.sql.catalog.v2.Catalogs.classKey;
+import static org.apache.spark.sql.catalog.v2.Catalogs.isOptionKey;
+import static org.apache.spark.sql.catalog.v2.Catalogs.optionKeyPrefix;
+import static scala.collection.JavaConverters.mapAsJavaMapConverter;
+
+@Private
+public class CatalogManager {
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/CatalogManager.java]: Why was it necessary to add this? It looks like convenience methods for tests.

I'd be fine adding this, but in a separate PR. I think adding this as well as tests for it over-complicates the addition of DROP TABLE.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/Catalogs.java [hunk 9]
```diff
@@ -35,6 +33,23 @@
   private Catalogs() {
   }
 
+  public static String classKey(String name) {
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/Catalogs.java]: Please remove unnecessary refactoring from this PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/DropTableStatement.scala [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+
+/**
+ * A DROP TABLE statement, as parsed from SQL.
+ */
+case class DropTableStatement(
+    tableName: Seq[String],
+    ifExists: Boolean,
+    isView: Boolean,
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/DropTableStatement.scala]: Combining tables and views in a single statement is an artifact of the parser. `DROP TABLE` and `DROP VIEW` are mutually exclusive, so I think we should use separate parsed commands for them.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 11]
```diff
@@ -649,8 +649,8 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
    * Create a [[DropTableCommand]] command.
    */
   override def visitDropTable(ctx: DropTableContext): LogicalPlan = withOrigin(ctx) {
-    DropTableCommand(
-      visitTableIdentifier(ctx.tableIdentifier),
+    sql.DropTableStatement(
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Now that this produces a plan that is defined in catalyst, it should be moved to `AstBuilder` in catalyst.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLParserSuite.scala [hunk 12]
```diff
@@ -906,59 +906,39 @@ class DDLParserSuite extends PlanTest with SharedSQLContext {
...
-        purge = false),
-      DropTableCommand(TableIdentifier("tab", None), ifExists = true, isView = false,
-        purge = false),
-      DropTableCommand(TableIdentifier("tab", None), ifExists = false, isView = false,
-        purge = true),
-      DropTableCommand(TableIdentifier("tab", None), ifExists = true, isView = false,
-        purge = true))
-
-    parsed.zip(expected).foreach { case (p, e) => comparePlans(p, e) }
+    Seq(
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLParserSuite.scala]: The approach for tests in SPARK-27108 was to avoid changing these tests so that we know that the behavior is exactly the same. Because the tests now require both parsing and running `DataSourceResolution`, the tests should be moved as-is to `PlanResolutionSuite`.

In addition, we also need to add new/updated parser tests. But because the parsing code can now be in Catalyst, the updated parser tests should be added to the catalyst `DDLParserSuite` instead of this one.

You can see how this was done for create and CTAS in https://github.com/apache/spark/pull/24029.

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java [hunk 13]
```diff
@@ -26,6 +26,7 @@
 import java.util.HashMap;
```
Comment (stagedataloss) [inline sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: HelloWorld

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java [hunk 14]
```diff
@@ -178,4 +179,17 @@ public double getDouble(String key, double defaultValue) {
   public Map<String, String> asCaseSensitiveMap() {
     return Collections.unmodifiableMap(original);
```
Comment (stagedataloss) [inline sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: HelloWorld

**Comment by jzhuge** [comment]
No longer depend on #24689.
Did a better job isolating the changes only needed for DropTable this time, so diff should be much more cleaner.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 15]
```diff
@@ -2195,4 +2195,22 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     }
   }
 
+  /**
+   * Create a [[sql.DropTableStatement]] command.
+   */
+  override def visitDropTable(ctx: DropTableContext): LogicalPlan = withOrigin(ctx) {
+    sql.DropTableStatement(
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Nit: I don't think `sql.` is required here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala [hunk 16]
```diff
@@ -83,6 +84,15 @@ case class DataSourceResolution(
             s"No catalog specified for table ${identifier.quoted} and no default catalog is set"))
           .asTableCatalog
       convertCTAS(catalog, identifier, create)
+
+    case DropTableStatement(CatalogObjectIdentifier(Some(catalog), ident), ifExists, _) =>
+      DropTable(catalog.asTableCatalog, ident, ifExists)
+
+    case DropTableStatement(AsTableIdentifier(tableName), ifExists, purge) =>
+      DropTableCommand(tableName, ifExists, false, purge)
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: Boolean arguments should use the named form: `isView = false`.

**Comment by rdblue** [comment]
+1 overall, just one minor style problem with boolean args to fix.

I think this is ready to go. You might also mention a couple of things in the PR description:
* This moves parsing of `DROP TABLE` into Catalyst and adds parsed plans, like #24029 did for create
* Like #24029, parsing tests for `DROP TABLE` have been moved to `PlanResolutionSuite` to validate existing behavior, and new tests for the catalyst parser have been added to Catalyst's `DDLParserSuite`

@cloud-fan, @dongjoon-hyun, could you review this DSv2 PR? I think it is ready to merge.

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @rdblue . Yep. I'll take a look, too~

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 15 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Right. Shall we use import, @jzhuge ?
```
-import org.apache.spark.sql.catalyst.plans.logical.sql.{CreateTableAsSelectStatement, CreateTableStatement}
+import org.apache.spark.sql.catalyst.plans.logical.sql.{CreateTableAsSelectStatement, CreateTableStatement, DropTableStatement, DropViewStatement}
```

Also, please remove `sql.` from the function comments, too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/DropViewStatement.scala [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+
+package org.apache.spark.sql.catalyst.plans.logical.sql
+
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+
+/**
+ * A DROP VIEW statement, as parsed from SQL.
+ */
+case class DropViewStatement(
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/DropViewStatement.scala]: Please mention this change in the PR description.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/DropViewStatement.scala [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+    tableName: Seq[String],
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/DropViewStatement.scala]: `tableName` -> `viewName`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 19]
```diff
@@ -499,6 +499,14 @@ object OverwritePartitionsDynamic {
   }
 }
 
+/**
+ * Drop a table.
+ */
+case class DropTable(
+    catalog: TableCatalog,
+    ident: Identifier,
+    ifExists: Boolean) extends Command
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Just to be sure, so, there is no `PURGE`. Did I understand correctly?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala [hunk 20]
```diff
@@ -83,6 +84,15 @@ case class DataSourceResolution(
...
+
+    case DropViewStatement(AsTableIdentifier(tableName), ifExists) =>
+      DropTableCommand(tableName, ifExists, true, false)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: ditto.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala [hunk 21]
```diff
@@ -438,4 +441,63 @@ class PlanResolutionSuite extends AnalysisTest {
...
+      DropTableCommand(tableIdent2, ifExists = false, isView = false, purge = false))
+    parseResolveCompare(s"DROP TABLE IF EXISTS $tableName2",
+      DropTableCommand(tableIdent2, ifExists = true, isView = false, purge = false))
+    parseResolveCompare(s"DROP TABLE $tableName2 PURGE",
+      DropTableCommand(tableIdent2, ifExists = false, isView = false, purge = true))
+    parseResolveCompare(s"DROP TABLE IF EXISTS $tableName2 PURGE",
+      DropTableCommand(tableIdent2, ifExists = true, isView = false, purge = true))
+  }
+
+  test("drop table v2") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala]: `drop table v2` -> `drop table for v2 catalog`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala [hunk 22]
```diff
@@ -438,4 +441,63 @@ class PlanResolutionSuite extends AnalysisTest {
...
+    parseResolveCompare(s"DROP VIEW $viewName2",
+      DropTableCommand(viewIdent2, ifExists = false, isView = true, purge = false))
+    parseResolveCompare(s"DROP VIEW IF EXISTS $viewName2",
+      DropTableCommand(viewIdent2, ifExists = true, isView = true, purge = false))
+  }
+
+  test("drop view v2") {
+    assertAnalysisError(
+      parseAndResolve("DROP VIEW testcat.db.view"),
+      Seq("unresolved operator 'DropViewStatement"),
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala]: We know the contexts, but for the general users, `unresolved operator ` message is misleading. Can we have a nicer exception message? This should be changed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala [hunk 23]
```diff
@@ -83,6 +84,15 @@ case class DataSourceResolution(
             s"No catalog specified for table ${identifier.quoted} and no default catalog is set"))
           .asTableCatalog
       convertCTAS(catalog, identifier, create)
+
+    case DropTableStatement(CatalogObjectIdentifier(Some(catalog), ident), ifExists, _) =>
+      DropTable(catalog.asTableCatalog, ident, ifExists)
+
+    case DropTableStatement(AsTableIdentifier(tableName), ifExists, purge) =>
+      DropTableCommand(tableName, ifExists, false, purge)
+
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: Please handle `DropViewStatement(CatalogObjectIdentifier(...))` case here to show a user-friendly `AnalysisException` message.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala [hunk 24]
```diff
@@ -23,12 +23,13 @@ import org.scalatest.BeforeAndAfter
...
 import org.apache.spark.sql.catalog.v2.Identifier
-import org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException
+import org.apache.spark.sql.catalyst.analysis.{NoSuchTableException, TableAlreadyExistsException}
 import org.apache.spark.sql.execution.datasources.v2.orc.OrcDataSourceV2
 import org.apache.spark.sql.test.SharedSQLContext
 import org.apache.spark.sql.types.{LongType, StringType, StructType}
 
-class DataSourceV2SQLSuite extends QueryTest with SharedSQLContext with BeforeAndAfter {
+class DataSourceV2SQLSuite
+  extends QueryTest with SharedSQLContext with BeforeAndAfter {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: Could you revert this unnecessary style change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 19 above)
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: That's right, DSv2 doesn't have a purge option.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala (hunk 22 above)
Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala]: We could implement a better rule in checkAnalysis for these, while there are not v2 implementations. Do you think this should be done in this PR, or could it be in a follow-up?

Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala]: Actually, I just saw your note about adding a rule for this with a better message. I think that's a good step for now.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala]: Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 15 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: +1, since we have the `Statement` postfix, there will be no confusion if we import these classes.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 25]
```diff
@@ -139,8 +139,8 @@ statement
         DROP (IF EXISTS)? partitionSpec (',' partitionSpec)*           #dropTablePartitions
     | ALTER TABLE tableIdentifier partitionSpec? SET locationSpec      #setTableLocation
     | ALTER TABLE tableIdentifier RECOVER PARTITIONS                   #recoverPartitions
-    | DROP TABLE (IF EXISTS)? tableIdentifier PURGE?                   #dropTable
-    | DROP VIEW (IF EXISTS)? tableIdentifier                           #dropTable
+    | DROP TABLE (IF EXISTS)? multipartIdentifier PURGE?               #dropTable
+    | DROP VIEW (IF EXISTS)? multipartIdentifier                       #dropView
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: do we have view support in DS v2? This is surprising to me, as I thought we may treat table/view differently in DS v2. Personally I think it's a bad idea that hive unifies table and view, it's weird to see the `sqlText` field in Hive `Table` class.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala (hunk 23 above)
Comment (jzhuge) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: Good idea!

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 25 above)
Comment (jzhuge) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Echo your sentiment towards Hive view handling. Separated visitView from visitTable and added DropViewStatement in preparation for better view handling, possibly even view catalog.

Comment (cloud-fan) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Are you going to add `DropView` operator in this PR? If not I think it's better to not add `DropViewStatement` in this PR. We can have a separated PR to add the view support.

I think we haven't reached a consensus that DROP VIEW is implemented via `TableCatalog.dropTable`.

Comment (jzhuge) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: I see, separate visitDropView makes it possible to delay moving DropView to catalyst. I am ok either way.

Comment (rdblue) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: I think since these were mixed together originally, it makes sense to move both statements to create parsed plans at the same time. We are going to need to do this eventually anyway.

Why change to use a separate `visitDropView`, but not also move it to Catalyst and use a parsed plan? Is there value in keeping it that way?

Comment (cloud-fan) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Yes Hive mixes DROP TABLE and DROP VIEW, but it doesn't mean we have to follow it in data source v2. At least this needs to be discussed.

It seems weird to me that we forward the DROP VIEW request to `TableCatalog` implementations, without official view support in data source v2.

Comment (rdblue) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: @cloud-fan, that's not what this PR does. This PR moves parsing into Catalyst by introducing a parsed plan, `DropViewStatement`, for `DROP VIEW`. We can still convert that to a `DropTable` in the analyzer if we choose to later, like we do for v1. (But I think it makes much more sense to use a `ViewCatalog`.)

The important thing is that we want to keep moving parsing code into Catalyst using parsed plans to separate the implementation (DataSource logical plans) from the parsing logic so that the parser isn't tightly coupled to DataSource v1.

Comment (cloud-fan) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: ah you are right! sorry I read the PR too fast

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by jzhuge** [comment]
Thanks @rdblue @dongjoon-hyun @cloud-fan !
